azure panic

Found an unchecked error case which can cause a panic. It's the array index on line 143

https://github.com/hashicorp/packer/blob/63f791570af9b11557816b324adb6b6a55926db1/builder/azure/arm/step_deploy_template.go#L138-L144

Seems like a simple enough fix, but I'm not familiar enough with this builder to say what should happen if the list has fewer than 2 or 3 elements.

cc @ArjenSchwarz

```
panic: runtime error: index out of range
2017/12/11 08:57:19 packer.exe: 	panic: runtime error: index out of range
2017/12/11 08:57:19 packer.exe:
2017/12/11 08:57:19 packer.exe: goroutine 193 [running]:
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).deleteImage(0xc042342f00, 0x0, 0x0, 0x0, 0x0, 0xc0422a6400, 0x11, 0xc0424586b0, 0x9)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:143 +0x402
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).(github.com/hashicorp/packer/builder/azure/arm.deleteImage)-fm(0x0, 0x0, 0x0, 0x0, 0xc0422a6400, 0x11, 0x9, 0x0)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:41 +0x70
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).Cleanup(0xc042342f00, 0x314b5c0, 0xc0422fa090)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:200 +0x889
2017/12/11 08:57:19 packer.exe: panic(0x1c704e0, 0x320e8e0)
2017/12/11 08:57:19 packer.exe: 	/usr/local/go/src/runtime/panic.go:491 +0x291
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).deleteImage(0xc042342f60, 0x0, 0x0, 0x0, 0x0, 0xc0422a6400, 0x11, 0xc0422b43a0, 0x9)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:143 +0x402
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).(github.com/hashicorp/packer/builder/azure/arm.deleteImage)-fm(0x0, 0x0, 0x0, 0x0, 0xc0422a6400, 0x11, 0x9, 0xc04228c5b0)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:41 +0x70
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*StepDeployTemplate).Cleanup(0xc042342f60, 0x314b5c0, 0xc0422fa090)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/step_deploy_template.go:200 +0x889
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep.(*BasicRunner).Run(0xc04214ac40, 0x314b5c0, 0xc0422fa090)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/vendor/github.com/mitchellh/multistep/basic_runner.go:82 +0x2e0
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/builder/azure/arm.(*Builder).Run(0xc04222fb00, 0x314e5a0, 0xc0423b7ee0, 0x3145840, 0xc0422b2970, 0x314ce80, 0xc0420043f8, 0x80000000018, 0x28, 0x28, ...)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/builder/azure/arm/builder.go:206 +0xdf8
2017/12/11 08:57:19 packer.exe: github.com/hashicorp/packer/packer/rpc.(*BuilderServer).Run(0xc04213d720, 0x1, 0xc0422060c0, 0x0, 0x0)
2017/12/11 08:57:19 packer.exe: 	/Users/mwhooker/go/src/github.com/hashicorp/packer/packer/rpc/builder.go:93 +0x1e4
2017/12/11 08:57:19 packer.exe: reflect.Value.call(0xc04215e7e0, 0xc0421460c0, 0x13, 0x20b0070, 0x4, 0xc042595f20, 0x3, 0x3, 0xc0423b7d01, 0x1a65100, ...)
2017/12/11 08:57:19 packer.exe: 	/usr/local/go/src/reflect/value.go:434 +0x90d
2017/12/11 08:57:19 packer.exe: reflect.Value.Call(0xc04215e7e0, 0xc0421460c0, 0x13, 0xc042595f20, 0x3, 0x3, 0xc0421600e0, 0x0, 0x0)
2017/12/11 08:57:19 packer.exe: 	/usr/local/go/src/reflect/value.go:302 +0xab
2017/12/11 08:57:19 packer.exe: net/rpc.(*service).call(0xc0421323c0, 0xc042142320, 0xc042288188, 0xc042464380, 0xc0421442c0, 0x1b2dc60, 0xc04220609c, 0x18a, 0x1ad5240, 0xc0422060c0, ...)
2017/12/11 08:57:19 packer.exe: 	/usr/local/go/src/net/rpc/server.go:381 +0x149
2017/12/11 08:57:19 packer.exe: created by net/rpc.(*Server).ServeCodec
2017/12/11 08:57:19 packer.exe: 	/usr/local/go/src/net/rpc/server.go:475 +0x372
```

--- a/builder/azure/arm/step_delete_os_disk.go
+++ b/builder/azure/arm/step_delete_os_disk.go
@@ -1,6 +1,7 @@
 package arm
 
 import (
+	"errors"
 	"fmt"
 	"net/url"
 	"strings"
@@ -80,11 +81,14 @@ func (s *StepDeleteOSDisk) Run(state multistep.StateBag) multistep.StepAction {
 	}
 
 	xs := strings.Split(u.Path, "/")
+	if len(xs) < 3 {
+		err = errors.New("Failed to parse OS Disk's VHD URI!")
+	} else {
+		var storageAccountName = xs[1]
+		var blobName = strings.Join(xs[2:], "/")
 
-	var storageAccountName = xs[1]
-	var blobName = strings.Join(xs[2:], "/")
-
-	err = s.delete(storageAccountName, blobName)
+		err = s.delete(storageAccountName, blobName)
+	}
 	return processStepResult(err, s.error, state)
 }
 

--- a/builder/azure/arm/step_delete_os_disk_test.go
+++ b/builder/azure/arm/step_delete_os_disk_test.go
@@ -104,6 +104,46 @@ func TestStepDeleteOSDiskShouldHandleComplexStorageContainerNames(t *testing.T) 
 	}
 }
 
+func TestStepDeleteOSDiskShouldFailIfVHDNameCannotBeURLParsed(t *testing.T) {
+	var testSubject = &StepDeleteOSDisk{
+		delete:        func(string, string) error { return nil },
+		say:           func(message string) {},
+		error:         func(e error) {},
+		deleteManaged: func(string, string) error { return nil },
+	}
+
+	// Invalid URL per https://golang.org/src/net/url/url_test.go
+	stateBag := DeleteTestStateBagStepDeleteOSDisk("http://[fe80::1%en0]/")
+
+	var result = testSubject.Run(stateBag)
+	if result != multistep.ActionHalt {
+		t.Fatalf("Expected the step to return 'ActionHalt', but got '%v'.", result)
+	}
+
+	if _, ok := stateBag.GetOk(constants.Error); ok == false {
+		t.Fatalf("Expected the step to not stateBag['%s'], but it was.", constants.Error)
+	}
+}
+func TestStepDeleteOSDiskShouldFailIfVHDNameIsTooShort(t *testing.T) {
+	var testSubject = &StepDeleteOSDisk{
+		delete:        func(string, string) error { return nil },
+		say:           func(message string) {},
+		error:         func(e error) {},
+		deleteManaged: func(string, string) error { return nil },
+	}
+
+	stateBag := DeleteTestStateBagStepDeleteOSDisk("storage.blob.core.windows.net/abc")
+
+	var result = testSubject.Run(stateBag)
+	if result != multistep.ActionHalt {
+		t.Fatalf("Expected the step to return 'ActionHalt', but got '%d'.", result)
+	}
+
+	if _, ok := stateBag.GetOk(constants.Error); ok == false {
+		t.Fatalf("Expected the step to not stateBag['%s'], but it was.", constants.Error)
+	}
+}
+
 func TestStepDeleteOSDiskShouldPassIfManagedDiskInTempResourceGroup(t *testing.T) {
 	var testSubject = &StepDeleteOSDisk{
 		delete: func(string, string) error { return nil },

--- a/builder/azure/arm/step_deploy_template.go
+++ b/builder/azure/arm/step_deploy_template.go
@@ -1,6 +1,7 @@
 package arm
 
 import (
+	"errors"
 	"fmt"
 	"net/url"
 	"strings"
@@ -140,6 +141,9 @@ func (s *StepDeployTemplate) deleteImage(imageType string, imageName string, res
 		return err
 	}
 	xs := strings.Split(u.Path, "/")
+	if len(xs) < 3 {
+		return errors.New("Unable to parse path of image " + imageName)
+	}
 	var storageAccountName = xs[1]
 	var blobName = strings.Join(xs[2:], "/")
 

--- a/builder/azure/arm/step_deploy_template_test.go
+++ b/builder/azure/arm/step_deploy_template_test.go
@@ -82,6 +82,31 @@ func TestStepDeployTemplateShouldTakeStepArgumentsFromStateBag(t *testing.T) {
 	}
 }
 
+func TestStepDeployTemplateDeleteImageShouldFailWhenImageUrlCannotBeParsed(t *testing.T) {
+	var testSubject = &StepDeployTemplate{
+		say:   func(message string) {},
+		error: func(e error) {},
+		name:  "--deployment-name--",
+	}
+	// Invalid URL per https://golang.org/src/net/url/url_test.go
+	err := testSubject.deleteImage("image", "http://[fe80::1%en0]/", "Unit Test: ResourceGroupName")
+	if err == nil {
+		t.Fatal("Expected a failure because of the failed image name")
+	}
+}
+
+func TestStepDeployTemplateDeleteImageShouldFailWithInvalidImage(t *testing.T) {
+	var testSubject = &StepDeployTemplate{
+		say:   func(message string) {},
+		error: func(e error) {},
+		name:  "--deployment-name--",
+	}
+	err := testSubject.deleteImage("image", "storage.blob.core.windows.net/abc", "Unit Test: ResourceGroupName")
+	if err == nil {
+		t.Fatal("Expected a failure because of the failed image name")
+	}
+}
+
 func createTestStateBagStepDeployTemplate() multistep.StateBag {
 	stateBag := new(multistep.BasicStateBag)
 

